### DISCUSSION TITLE
Docker Container

### DISCUSSION THREAD

**Discussion by josephcooley** [maforget/ComicRackCE#21] [Ideas]
I know this is still very new and super early in the development, but how difficult would it be to create a docker container for this eventually?  I never really used Comic Rack as a reader, but the sorting and renaming server side portion of CR is amazing!

**Comment by maforget** [comment]
Someone already did for the original. https://github.com/pezhore/comicrack-docker, https://www.reddit.com/r/comicrackusers/comments/17rlgfb/dockerizing_comicrack_need_help_understanding/, https://www.reddit.com/r/comicrackusers/comments/17vh35d/comicrack_docker_container_release_v001apre/

But I honestly have 0 intention of doing so. Last time I tried it on a Windows host, I found out that I couldn't bind an address to an outside address. Which seems worthless to me. If I have to install it and use it on my machine might as well use it directly.

If you are talking about the Remote Library, it's currently not working (#12). And you would need to have a client installed anyway. You could host the files on a network server and organize them via the local client.

**Reply by Nadiar** [reply]
It would be ideal to use a docker container either via wine (like pezhore did) or dotnet, because people could just run it on their Synology or other NAS storage device.  I understand a much lower priority though, as the process order needs to be replacing the plugin system, then getting a headless system working.

**Reply by maforget** [reply]
I still don't get the appeal of a Docker container, what is the usage here? Why docker? Make it make sense. You want to use it as a server and use the library share feature? Because I won't work on something if I don't use it or see any uses to it. If someone wants to pitch in, by all means. And if it's just because I like docker than just no.

The only reason I would install a software like this on a NAS would be to use has a server. But using on a very under powered device isn't really optimal. Much better to just leave the files on the NAS and connect there from a computer. If you want a server, than I kind of get it. But still there are alternatives really more suited to this like Komga.

For having a native linux version, even if using the latest .NET that can build native linux versions, there is a shit ton of code that uses Win32 API in here. So I don't know if it would be possible. 

Although a headless version is probably doable it would require a way to control it. Right now a lot of that already exists in command lines (check the wiki) or the ComicRack.ini. Really the only thing needed would be to just start the server sharing, and that requires some config changes that only exists in the UI. Even then that also uses some Windows Communication Foundation to communicate, that is looks like it may be a thing on Linux.

And using Docker on a windows host is kinda pointless has I see it, so with the app firmly in the Windows Eco-system, the only resort are thing like Wine. I don't use linux really, so I ma not really the person to port that and really don't see what I could change to the ComicRack program that would help that.

**Comment by maforget** [comment]
Just fixed #12

**Comment by josephcooley** [comment]
That is the the exact use case that I suggested this for.  I have Komga and
use it to read comics.  But there is nothing that comes close to comic rack
for updating and managing the metadata and file/folder structure.  I used
CR this way on my old windows server back when CR was still getting
updates.  Having everything on a single server is attractive to me as a non
enterprise user that have everything on a truenas core machine.

**Reply by maforget** [reply]
I guess you are continuing the discussions from the other thread. I guess you are replying from email, I think you have to do it from the site to reply to the thread.

My thinking was install Komga on the NAS that can be access to all the network to read comics, but still have ComicRack on the computer that access the files from the NAS, and manages them. So ComicRack on the computer with bigger hardware so faster converting, etc. And Komga on the NAS just for access.

My question remains but why Docker? Does the NAS software only accept dockers container? I usually see Docker usage with programs that run server that use python or NodeJs, where the repo can be ran directly from the source. I see it as being easier to have separate environments when running a lot of different servers, that might use different python version, dependencies version, etc. And updating from the source is easy with updating to the new version when it's in. I get this usage of Docker.

But for a windows only app, that at the very least requires a UI for the initial setup. And the server part isn't the main feature, like Jellyfin/Emby. Where code needs to be compiled and all dependencies are shipped with it, the need for Docker really decreases. 

So what exactly is required for ComicRack to make it compatible? Have a headless server like in #22, for easy access to the library sharing. What about adding a command line so it starts in the background with the server setting passed by the command line. Would that be enough? Does the library feature work under wine with the container done by @pezhore? Is it that feature that is required to be made usable on linux? What exactly do I need to change/fix so that the container is doable? 

So the attraction of a Docker Container is exclusively for the library sharing feature or is there another?

**Reply by josephcooley** [reply]
The attraction for the docker container would be the library sharing, and managing the library.  The files are all on my NAS to begin with so to me it makes sense to have the program managing them on the same server.  I understand what you're saying though, maybe its just how I run my home network.  I have a main NAS server that I store all my data, and programs on (Plex, Tautuli, Komga, jackett, etc.) that is a rack mounted headless server, its always on and always running.  All other computers in the network are pure clients, they only pull data from the NAS, nothing but the basics installed on them.  So I could reformat/ lose/ set on fire any computer in the network and my life would be unaffected other than the loss.  Installing ComicRack on a computer other than the NAS breaks that.  That is my use case, and why I made the suggestion, it might not be for everyone.  Docker allows me to do that easily.  The other option to achieve this is to create a windows VM to run comic rack which is overkill.

**Reply by maforget** [reply]
But to use the library sharing in ComicRack still requires to install it on the client to access it? So are you remoting to the server? Than the library function doesn't apply to you and a headless server wouldn't be the solution?

**Reply by maforget** [reply]
Also what prevents you from installing the program with Wine on the server directly? Does it absolutely have to be docker?

**Reply by josephcooley** [reply]
TrueNAS doesn't run a traditional Linux OS, as far as i know there is no way to install actual Linux programs like wine.  It  runs the TrueNAS NAS operating system and then has the ability to run docker images.  It can also spin up VMs but again a whole VM for CR seems like too much.  The main thing I wanted that i cant find another solution for is the  file/folder/metadata management.  Comic Rack is unequaled in that category.  The server functionality is just nice but not the primary reason that I'd want it.  I guess it would need a web interface to replicate the windows gui, which now that I'm saying it out loud seems like a bigger ask than I originally thought.

**Reply by Salvora** [reply]
I also agree with @maforget, currently there is no point of having it run on docker. Maybe in the long run, when it has moved to .Net and you can use it headless. Then a dockerfile can be generated.

**Reply by Nadiar** [reply]
I would agree the priority is to get everything up to the most recent .net before even considering spending time trying to get this containerized.  I also think the idea of running headless in a container isn't needed. You could do VNC or Guacamole to get GUI access.
